openmsupply/open-msupply
# Refactor authentication

Fixes #1019 (🤞)

# 👩🏻‍💻 What does this PR do? 
 
I was hoping to implement one simple fix and resolve the occasionally logged out issue. This is slightly more involved that I'd hoped - but it boils down to one main change: the refresh token handling in f/e was not actually operating correctly.

The summary is: when you login an authentication cookie is saved to your browser. this lasts for 60 minutes after which you are logged out. If you make a graphQL request then the cookie expiry is updated, so that you get logged out 60 minutes after your last API interaction.

Partly this comes down to a problem where the state update was happening outside of the component and raising this error:

`Can't perform a React state update on an unmounted component`

the reasons for which were slightly complicated - partly it is because the `AuthContext` component was doing too many things, and had to re-render often. The internal state i.e. when the token changed, causes a render as you'd expect. Many of the hooks were using `useMutation` which requires internal state changes, which then causes the `AuthContext` to re-render. etc. 

I've lifted some functionality out to the `useLogin` hook to isolate those changes. The refresh token has been taken out and implemented differently, as that was also causing issues.

An outline of the auth process is now:

[![](https://mermaid.ink/img/pako:eNptU8Fy2jAQ_ZUdnZwhzPTUA4d2CBTqBhIScsMcFHuJNdgSkVZtqOHfu5acjMPUJ41231u999aNyE2BYiRerDyU8DTNNPA3ThbmRekrGA6_wU2yRgLv0ILUBTgyFkFpGHsqJ0YTvtFVhDn_HHkCGki6vYuVm8A0CUySccaqv5KU0VCiLJiZ2-ev1We6SQBNP0CQG7NX2FWnofpwWQV8Oyh7BFI1Ahl4RqikIxivUpB5js7F0gC-foGaNUY21EU8_GilG09R-yyZVCjtf8b3tXJ7X-wsQOcdFK011l2O-ZnMW_TDAiy-enTdvLRJHd_sLLryyexRg7HgjjpP9c4AN9rj93NkSAPgpM0JHpN1CKWnr5v3a9NLacuIIdxuNp1FA_anHTGI4Q5itNtthN4G_kXTrElaCqT23M1exNn4m9_TmugJT7BslIvuY_H-yGVsPKI7wezTVfvuuxYhq9pwPhfAux7wvm37CDGYAH8UlbwzMUP3jrrvoVbJY_QxquwMWcWVzrS4Fiyolqrg7W_aYiaoxBozMeJjgTvpK8pEps_cyvmbNccgRmQ9Xgt_KCThVEnegVqMdrJyfIuFYgeX8Y8KP9b5H-yDFLQ?type=png)](https://mermaid.live/edit#pako:eNptU8Fy2jAQ_ZUdnZwhzPTUA4d2CBTqBhIScsMcFHuJNdgSkVZtqOHfu5acjMPUJ41231u999aNyE2BYiRerDyU8DTNNPA3ThbmRekrGA6_wU2yRgLv0ILUBTgyFkFpGHsqJ0YTvtFVhDn_HHkCGki6vYuVm8A0CUySccaqv5KU0VCiLJiZ2-ev1We6SQBNP0CQG7NX2FWnofpwWQV8Oyh7BFI1Ahl4RqikIxivUpB5js7F0gC-foGaNUY21EU8_GilG09R-yyZVCjtf8b3tXJ7X-wsQOcdFK011l2O-ZnMW_TDAiy-enTdvLRJHd_sLLryyexRg7HgjjpP9c4AN9rj93NkSAPgpM0JHpN1CKWnr5v3a9NLacuIIdxuNp1FA_anHTGI4Q5itNtthN4G_kXTrElaCqT23M1exNn4m9_TmugJT7BslIvuY_H-yGVsPKI7wezTVfvuuxYhq9pwPhfAux7wvm37CDGYAH8UlbwzMUP3jrrvoVbJY_QxquwMWcWVzrS4Fiyolqrg7W_aYiaoxBozMeJjgTvpK8pEps_cyvmbNccgRmQ9Xgt_KCThVEnegVqMdrJyfIuFYgeX8Y8KP9b5H-yDFLQ)

In addition - I noticed that the `me` query was being sent twice in quick succession - because there was a mutation calling `me` as well as a query. To reduce the server calls I've updated the query cache with the result of the mutation call ( as this happens first ) but it doesn't appear to have helped.

# 🧪 How has/should this change been tested? 

- You can set the login timeout to something less than 60 minutes, login and wait for the system to time out and log you out
- Login, and check the auth cookie expiry. Interact with the site and see that the expiry does not change
- Login, and check the auth cookie expiry. Interact with the site when there is less than a minute until the cookie expires. see that the expiry is extended 
- leave the site alone and see if you are logged out before you should be

## 💌 Any notes for the reviewer?

Good luck!

## 📃 Documentation

_No user facing changes_
